pkg/cloudprovider/utils.go

```diff
@@ -0,0 +1,99 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cloudprovider
+
+import (
+	"fmt"
+	"net"
+	"strings"
+)
+
+const (
+	defaultLBSourceRange = "0.0.0.0/0"
+)
+
+// TODO: Should I move this into utils/sets ?  Even though it isn't generated?
+type IPNetSet map[string]*net.IPNet
+
+func ParseIPNetSet(specs []string) (IPNetSet, error) {
+	ipnetset := make(IPNetSet)
+	for _, spec := range specs {
+		spec = strings.TrimSpace(spec)
+		_, ipnet, err := net.ParseCIDR(spec)
+		if err != nil {
+			return nil, err
+		}
+		k := ipnet.String() // In case of normalization
+		ipnetset[k] = ipnet
+	}
+	return ipnetset, nil
+}
+
+// StringSlice returns a []string with the String representation of each element in the set.
+// Order is undefined.
+func (s IPNetSet) StringSlice() []string {
+	a := make([]string, 0, len(s))
+	for k := range s {
+		a = append(a, k)
+	}
+	return a
+}
+
+// Equal checks if to IPNetSets are equal (ignoring order)
+func (l IPNetSet) Equal(r IPNetSet) bool {
+	if len(l) != len(r) {
+		return false
+	}
+
+	for k := range l {
+		_, found := r[k]
+		if !found {
+			return false
+		}
+	}
+	return true
+}
+
+// Len returns the size of the set.
+func (s IPNetSet) Len() int {
+	return len(s)
+}
+
+// GetSourceRangeAnnotations verifies and parses the LBAnnotationAllowSourceRange annotation from a service,
+// extracting the source ranges to allow, and if not present returns a default (allow-all) value.
+func GetSourceRangeAnnotations(annotation ServiceAnnotation) (IPNetSet, error) {
+	val := annotation[LBAnnotationAllowSourceRange]
+	val = strings.TrimSpace(val)
+	if val == "" {
+		val = defaultLBSourceRange
+	}
+	specs := strings.Split(val, ",")
+	ipnets, err := ParseIPNetSet(specs)
+	if err != nil {
+		return nil, fmt.Errorf("Service annotation %s:%s is not valid. Expecting a comma-separated list of source IP ranges. For example, 10.0.0.0/24,192.168.2.0/24", LBAnnotationAllowSourceRange, val)
+	}
+	return ipnets, nil
+}
+
+// IsAllowAll checks whether the IPNetSet contains the default allow-all policy
+func IsAllowAll(ipnets IPNetSet) bool {
+	if ipnets.Len() != 1 {
+		return false
+	}
+	s := ipnets.StringSlice()
+	return s[0] == "0.0.0.0/0"
```

Comment:
length == 1 here.  But I'm going to have to fix that because of the "weird" edge case.  The real goal here was not to break OpenStack if the user specified "0.0.0.0/0" (the one policy OpenStack can support, and also the default)